compiler/src/dmd/dsymbolsem.d

```diff
@@ -3050,7 +3050,7 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
         if (sc.flags & SCOPE.compile)
             funcdecl.isCompileTimeOnly = true; // don't emit code for this function
 
-        funcdecl._linkage = sc.linkage;
+        funcdecl._linkage = (funcdecl.isCrtCtor || funcdecl.isCrtDtor) ? LINK.c : sc.linkage;
```

Comment:
`extern(C)` is problematic wrt. name clashes due to no mangling. It's the main reason LDC hasn't rejected `extern(D)` CRT c/dtors - they are ABI-compatible anyway.

Reply:
I know, but them being ABI-compatible is happenstance. I figure that to be pedantically correct, they should use C linkage. I can add a "best practice" note in the documentation that adding a module name prefix to it would be good.

Reply:
It's happenstance for Win32 only, otherwise the D ABI spec defines it as C-compatible. IMO, requiring the user to unique the names is clearly a PITA and in no way justifies this pedantic-ness about the linkage.

Reply:
Some evidence: https://github.com/dlang/dmd/blob/master/druntime/src/core/memory.d#L186-L206

Reply:
A technique I usually use:
```D
alias SigHandlerT = extern(C) void function (int sig) nothrow;

/// Returns a signal handler
/// This routine is there solely to ensure the function has a mangled name,
/// and doesn't accidentally conflict with other code.
private SigHandlerT getSignalHandler () @safe pure nothrow @nogc
{
    extern(C) void signalHandler (int signal) nothrow
    { }
    return &signalHandler;
}
```

Reply:
I fixed it so it has C linkage, but D mangling. That should make everyone happy :-)

Reply:
That will help me. It's a bit of a concern I've had that I might have to limit myself to just one to avoid name clashes.
But yeah it's going to be a bit of a surprise so will need documenting.

Reply:
> I fixed it so it has C linkage, but D mangling.

That's fine for the default, but when I declare this:

```D
pragma(crt_constructor) extern(C) myCrtCtor();
```

I wouldn't expect this to get D mangling, but it's what this PR would do currently.

Reply:
Good point, and also clearly visible from Walter's `pragma(mangle)` stuff to keep it backwards-compatible. Which clearly shows that this might easily break existing code.

Reply:
> I wouldn't expect this to get D mangling, but it's what this PR would do currently.

There are other existing cases where this happens, you can see it in the function where I made this change. If crt_constructors are used as intended, one would never see the name mangling.

> and also clearly visible from Walter's pragma(mangle) stuff to keep it backwards-compatible.

What was done in the gc code in this PR was the original code was clearly the wrong way to do things, it just happened to work. To minimize the diffs in this PR, I just made it work. The better solution is to put the constructors in their own module, so the name monkey business would not be necessary, it would just work.

@kinke it did not break existing code with the C mangling it did before, but you asked to change it. What else can I do? I cannot fix the ModuleInfo bugs many are complaining about without changing *something* in the way crt_constructors work. What do you suggest?

Reply:
> but you asked to change it.

Nope. I'm just advocating for leaving the linkage alone - as LDC did from the start, or for a very long time at least. No ModuleInfo bugs require an `extern(C)` linkage for CRT c/dtors.

Reply:
I.e., keeping up https://dlang.org/changelog/2.101.0.html#dmd.crt_constructor_signature.

Reply:
Leaving it alone means it has D linkage, which is *assumed* to be the same as C linkage. It is not, even if it is the same for some platforms. It is a hack to rely on them being the same.

Thanks for the reference. This is recent, and documenting what is going on is not fixing it. C linkage *is* required because the C runtime library *expects* it. We cannot define it away.

> No ModuleInfo bugs require an extern(C) linkage for CRT c/dtors.

Right. It's the C runtime library that requires it. Fixing the ModuleInfo problem requires fixing the problems with crt_constructors so they are specified properly.

Reply:
> Fixing the ModuleInfo problem requires fixing the problems with crt_constructors so they are specified properly.

Oh, I see where you are going with this, it's for the patch function!